rasterio/_io.pyx

```diff
@@ -914,6 +918,29 @@ cdef class DatasetWriterBase(DatasetReaderBase):
         name_b = path.encode('utf-8')
         fname = name_b
 
+        # Process dataset opening options.
+        for k, v in kwargs.items():
+            # Skip items that are definitely *not* valid driver
+            # options.
+            if k.lower() in ['affine']:
+                continue
+
+            k, v = k.upper(), str(v).upper()
+
+            # Guard against block size that exceed image size.
+            if k == 'BLOCKXSIZE' and int(v) > width:
+                raise ValueError("blockxsize exceeds raster width.")
+            if k == 'BLOCKYSIZE' and int(v) > height:
+                raise ValueError("blockysize exceeds raster height.")
+
+            key_b = k.encode('utf-8')
+            val_b = v.encode('utf-8')
+            key_c = key_b
```

Comment:
What's the purpose of defining `key_b` and then assigning to `key_c`? I see a handful of other examples like this in the changes here so curious what necessitates it.

Reply:
@jdmcbr It converts Python bytes to a new, fully independent C string.  This pattern is useful for making strings usable inside `with nogil:` too.

Reply:
Though now that you mentioned it, I removed some code and defs in favor of casting in the functions calls like `CSLAddString(drivers, <const char *>string_b)`.

Reply:
@jdmcbr To clarify, (AFAIK) casting inline is not supported in `with nogil:`  Everything referenced inside `with nogil:` must already be a C object because the GIL has already been released by the time the Python object must be cast to a C object.

Reply:
@geowurster @sgillies Thanks for the explanation, I like the change to use `CSLAddString`.